### ISSUE TITLE
distsql: Error "unknown prefix of the encoded byte slice" while running query

### ISSUE THREAD

**Issue by dt** [cockroachdb/cockroach#27648]
Running the following query against a table with a id column that uses a `BYTES` type to store UUIDs (it predates UUID support), an `INT` count column and a `TIMESTAMP` column.

```
root@:26257/registration> select sum(count), id, date_trunc('month', timestamp) from mytbl where timestamp > '2018-07-01' group by id, date_trunc('month', timestamp) having sum(count) > 1000;
pq: unknown prefix of the encoded byte slice: "\x12acbebc58-7d0d-4f1e-b5c7-2b531ff93c51\x00\x01"
root@:26257/registration> select sum(count), id, date_trunc('month', timestamp) from mytbl where timestamp > '2018-07-01' group by id, date_trunc('month', timestamp) having sum(count) > 1000;
pq: unknown prefix of the encoded byte slice: "\x125dc13b7b-d513-4038-87db-c1e3c8806175\x00\x01"
root@:26257/registration>
```

confusingly, the byte slice in that error looks a lot like the content of the UUID column, but that error message only appears to be generated by float and decimal decoding?

**Comment by dt** [comment]
@jordanlewis @knz not sure where this should go for triage. I have a reliable repro on an internal cluster that can be discussed offline.

**Comment by knz** [comment]
I don't know who are the current experts with encoding. I am not paged in on that stuff currently. 
Based on `git log` I'd suggest @mjibson and @justinj.

**Comment by solongordon** [comment]
Doesn't look like this is actually an encoding issue but rather a distsql planning or execution one. Something about the order of the output columns combined with the `HAVING` clause is messing things up. I have a minimal repro, which fails on `release-2.0` but not `master`. I'll dig in further.

```
root@:26257/d> create table t (k text primary key);
root@:26257/d> insert into t values ('a'), ('b'), ('c');
root@:26257/d> alter table t split at values ('b'), ('c');
root@:26257/d> alter table t experimental_relocate values
  (array[1], 'a'),
  (array[2], 'b'),
  (array[3], 'c');
root@:26257/d> select count(*), k from t group by k having count(*) > 0;
pq: insufficient bytes to decode uvarint value: "a\x00\x01"
```

**Comment by jordanlewis** [comment]
Huh. I wonder if we fixed this on master and didn't backport.

**Comment by solongordon** [comment]
This is a bug in the DistSQL physical planning logic for de-duplicating aggregations. The HAVING clause causes the logical planner to plan two `COUNT(*)` aggregations, which distsql deduplicates into one. However, this is not taken into account for the aggregator's `OutputTypes`, so there is a mismatch between the output rows and the reported output types.

It looks like this was "fixed" by https://github.com/cockroachdb/cockroach/pull/24478 because the deduplication now happens during logical planning. (The optimizer also has this logic.) But the sketchy physical planning logic remains, and there might still be some way to trigger bad behavior.

I'm going to look into how straightforward it is to fix the physical planner, or whether it's better to just backport https://github.com/cockroachdb/cockroach/pull/24478 and call it a day.

**Comment by solongordon** [comment]
Fixed by backport: #28625.
